### ISSUE TITLE
Make our password hint progressive

### ISSUE THREAD

**Issue by ryanfeeley** [mozilla/fxa-content-server#3731]
![screenshot 2016-05-12 16 39 32](https://cloud.githubusercontent.com/assets/68213/15229690/36aec66c-1860-11e6-8eee-ec5c6d39ca28.png)
Currently the only password advice we give registering users is "Must be at least 8 characters". This is inadequate the users who plan to sync their browsing data, including passwords, so we should discuss  making improvements to this experience.

We currently run a Bloom filter to gauge password complexity, but do not expose it in the interface.

What if we changed our our hint (i.e. "Must be at least 8 characters") with every keystroke, until the password was of adequate strength? We could also use this as an opportunity to increase the understanding of how sync works.

Examples on the web:
* https://howsecureismypassword.net/
* http://www.trypap.com/

And what 1Password does:
<img width="762" alt="screen shot 2016-05-11 at 11 48 00 am" src="https://cloud.githubusercontent.com/assets/68213/15229364/b93580f0-185e-11e6-8dd7-c9ea9a1c3481.png">
<img width="762" alt="screen shot 2016-05-11 at 11 48 04 am" src="https://cloud.githubusercontent.com/assets/68213/15229369/b9392822-185e-11e6-8d1e-7d8ba27a5dc3.png">
<img width="762" alt="screen shot 2016-05-11 at 11 48 12 am" src="https://cloud.githubusercontent.com/assets/68213/15229368/b9391ddc-185e-11e6-9947-a0e4d311d067.png">
<img width="762" alt="screen shot 2016-05-11 at 11 48 17 am" src="https://cloud.githubusercontent.com/assets/68213/15229365/b9365b2e-185e-11e6-87b9-4e89d674f33c.png">

**Comment by TDA** [comment]
+1 to this. I am willing to do take this up :)

**Comment by ryanfeeley** [comment]
Some information we might want to get across:

- 8 characters are required to create an account
- A strong password is recommended if you plan to sync passwords
- A short phrase is usually a strong password – it is easy to type and also remember
- If you reset your password, you will lose any Sync data that is not on one of your devices
- Our Sync servers are designed as go-between your devices, not a standalone backup

**Comment by Verdi** [comment]
I like this idea. If the text changed with each keystroke, the default prompt could be something that lets people know that this password protects all the passwords they've saved in Firefox.

**Comment by ryanfeeley** [comment]
Would be interesting to preface every line with "This password"…

**Comment by rfk** [comment]
I'm broadly 👍 to this idea, and I wonder if we can pare it right down to a minimal version to try it out quickly in practice.  What if a first version of this just had a handful of the most important states:

* Default: `"Pick a strong password, this is used to encrypt all your browser data"` or whatever wordsmithery we can come up with for the initial prompt
* When they start typing, transition to  `"Must be at least 8 characters"`
* When they exceed that limit, check against the list of common passwords and show `"That's a very common password, choose something more unique"`

Then if we like it and it seems to be valuable, we can iterate by adding additional checks and messaging.

**Comment by TDA** [comment]
I am very much inclined towards the approach @rfk specifies. We should get it running to a minimal version before we decide to add more functionality.
If we get to know that the basic version is useful, we could then go ahead and even test out the password playground, by letting people type something and generate a password from that.

**Comment by ryanfeeley** [comment]
Some wordsmithed copy (I'll get @mozmatej to sanity check it)

Default (focused, empty):
* A strong, unique password will keep your Firefox data safe from prying eyes. [Need inspiration](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)?

With characters entered:
* 8 characters minimum, but longer if using Sync for passwords. [Need inspiration](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)?

With a weak password:
* That password is overly common and not recommended; please choose something more unique.  [Need inspiration](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)?

**Comment by ryanfeeley** [comment]
Some wordsmithed copy (I'll get @mozmatej to sanity check it)

Default (focused, empty):
* A strong, unique password will keep your Firefox data safe from prying eyes. [Need inspiration](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)?

With characters entered:
* 8 characters minimum, but longer if using Sync for passwords. [Need inspiration](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)?

With a weak password:
* That password is overly common and not recommended; please choose something more unique.  [Need inspiration](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)?

**Comment by vladikoff** [comment]
Needs feature doc

**Comment by ryanfeeley** [comment]
Assigned to myself for feature doc.

**Comment by vladikoff** [comment]
👍

**Comment by ryanfeeley** [comment]
Also realizing that we don't want people to include local attackers with "prying eyes". Damn, this is going to be hard.

**Comment by ryanfeeley** [comment]
The tricky thing will be to be compatible with our signup.signin flow.

**Comment by rfk** [comment]
It's not obvious from the comment history here - what is this currently blocked on?

**Comment by TDA** [comment]
@rfk
We need the link for `More info`

**Comment by rfk** [comment]
From https://github.com/mozilla/fxa/pull/150/files#diff-c00b54226abc4785bd4d2f483d4e24b1R54 I see the following note:
```
"More info" will link to a modal dialog containing concise instructions on creating a strong password based on [this SUMO article](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe)
```

Which suggests we may actually have to host our own copy, not just an outgoing link.

Could we consider landing without the "more info" link and doing it as a separate follow-up bug?  The new functionality seems to provide significant value to me even without the extra context.  (We would, of course, have to practice the whole "not losing the follow-up bugs" thing that @ryanfeeley has noted in the past).

**Comment by ryanfeeley** [comment]
@rfk I think it's a constraint on mobile where the web views cannot link externally :(

**Comment by ryanfeeley** [comment]
I suggest we copy the contents from [this page](https://support.mozilla.org/en-US/kb/create-secure-passwords-keep-your-identity-safe) and present it on a page much like our Privacy Notice / Terms of Service pages.
![screenshot 2016-06-29 13 24 27](https://cloud.githubusercontent.com/assets/68213/16461940/fde3993a-3dfc-11e6-940e-c507e49a071a.png)

Can we include the video?
We don't need the table of contents or any contents beyond this line: "Take a moment to think of a phrase that's meaningful to you. Use that phrase to create a secure password that you can customize for each website you visit."

**Comment by TDA** [comment]
So is this the ideal course of action:
* Create a new template with the contents of the SuMo article (without ToC and other stuff)
* Add this to existing routes.
* Link to this document using the `More Info` link.

Question: What should be the name of this document? 
@shane-tomlinson @rfk

**Comment by rfk** [comment]
I really don't want to grow a clone of SUMO under our own URL space...but if there's no other way to show good messaging to mobile users, I guess we'll need to display it ourselves.  How do we currently handle localization of the legal and terms pages, and can we do the same for this copy?

Paralleling https://accounts.firefox.com/legal/terms and https://accounts.firefox.com/legal/privacy, perhaps we can start a https://accounts.firefox.com/support namespace and put this copy into it at https://accounts.firefox.com/support/create-a-secure-password

@ryanfeeley do you think we need to say something about *why* the security of the password here is so important, i.e. something to explain the "longer if you're syncing passwords" part?

**Comment by TDA** [comment]
From mtg:
`ship it without more info link, make the "More info " page a separate PR.`
